### ISSUE TITLE
FP16 support

### ISSUE THREAD

**Issue by Randl** [tiny-dnn/tiny-dnn#355]
Lower precious numbers and especially [half-precious floats](https://en.wikipedia.org/wiki/Half-precision_floating-point_format) became popular recently in Deep Learning community, since they allow faster and less memory-intensive learning with [very small loss of accuracy](https://arxiv.org/abs/1502.02551). 

Latest GPUs show [double speedup](https://images.nvidia.com/content/pdf/tesla/whitepaper/pascal-architecture-whitepaper.pdf) for FP16 over FP32. 
Even though FP16 isn't supported natively in x86 processors, [F16C](https://en.wikipedia.org/wiki/F16C) instruction set allows efficiently convert FP16 to FP32, still [benefiting](https://software.intel.com/en-us/articles/performance-benefits-of-half-precision-floats) from half memory usage.

I suggest to use `uint16_least_t` to store FP16 values.

**Comment by Randl** [comment]
Here is my version of half class:
https://gist.github.com/Randl/169004c2377fef7db28a418cc72e330c

**Comment by edgarriba** [comment]
@Randl looks good! Could you provide some tests to see how does it work?

**Comment by bhack** [comment]
This is what one of our upstream (CPU) backend recently added https://github.com/Maratyszcza/NNPACK/blob/master/include/nnpack/fp16.h

**Comment by bhack** [comment]
@naibaf7 Any news on https://github.com/naibaf7/libdnn/issues/7#issuecomment-234569718?

**Comment by Randl** [comment]
Intuitively, converting to float and performing operations natively should be faster. The only problem is that it requires AVX.

**Comment by bhack** [comment]
@Maratyszcza have you implemented in this way for having more portability?

**Comment by Maratyszcza** [comment]
@bhack I have an implementation of fully-connected inference with FP16 weights, albeit it is not merged into public repo. No plans for other FP16 options so far.

**Comment by bhack** [comment]
@Maratyszcza I was talking of the conversion implementation. Why you are not using SIMD instruction directly?

**Comment by Maratyszcza** [comment]
@bhack This is a reference implementation for platforms which do not have native FP16->FP32 conversion. Besides, it is branch-free, fast, and SIMD-friendly.

**Comment by Maratyszcza** [comment]
Here is a PSIMD version, if you are interested: https://gist.github.com/Maratyszcza/0d7f79cb618adf683f9bdeeb9a268bfe

**Comment by Randl** [comment]
The code [runs](https://gist.github.com/Randl/169004c2377fef7db28a418cc72e330c) and passes [tests](https://gist.github.com/Randl/3aea2f9a406f022931a64d7ec9242a8c).

**Comment by bhack** [comment]
@edgar @naibaf7 What do you thin of FP16 section of https://github.com/CNugteren/CLBlast/blob/master/README.md? What we need (our side) to handle for FP16 in libdnn?

**Comment by CNugteren** [comment]
CLBlast's kernels do indeed support FP16  (although not always fully working on the devices I tested on - still some work to do). From the OpenCL host side there is nothing special required, `cl_half` is already a suitable data-type (16-bit integer on the host side).

If you want to do arithmetic on the C++ side outside of OpenCL, then you'll have to convert back and forth to single-precision floating-point. In CLBlast I provide some [simple utility functions](https://github.com/CNugteren/CLBlast/blob/master/include/clblast_half.h). For a more complete version you can look at the [half library](http://half.sourceforge.net/).

**Comment by bhack** [comment]
Just to notify the isocpp proposal in WG14 group: http://www.open-std.org/jtc1/sc22/wg14/www/docs/n2016.pdf

**Comment by Randl** [comment]
https://github.com/Randl/tiny-cnn/tree/half

**Comment by Maratyszcza** [comment]
@Randl Why do you want to cast to half-precision after every operation? It is slower than single-precision computation (because of conversion overhead), less accurate than single-precision computation (because of loss of precision on down-conversion) and even less accuracy than half-precision computation (because of double rounding). I think the proper way to handle half-precision on systems that do not support half-precision computations is to do all computations in single-precision, and only convert the final result to half.

**Comment by Randl** [comment]
@Maratyszcza main purpose is gpu support and vectorization which will be done. Doing computations in half precious is slower

**Comment by beru** [comment]
FP16 support is appealing since it halves data size and required bandwidth.
But I think there're several obstacles.

- Std C++ language spec doesn't support Short Float type, maybe from C++20. https://github.com/opencv/opencv/issues/8428#issuecomment-291758060 
- Currently, tiny-dnn's `float_t` type can be either `float` or `double`. If possible, tensor_t storage should support FP16 type with something like typedef of `uint16_t` but internal backend implementation doesn't support FP16 computation at all so overhaul change with type conversion will probablly be needed. 
- It looks like F16C (FP16C) instructions have high latencies and it seems like current (and even 
 future) x86/x64 processors seem not to support native FP16 computation and that makes float16 path less appealing for CPU backends.
  https://en.wikipedia.org/wiki/AVX-512#Floating_point_decomposition
  https://software.intel.com/en-us/articles/performance-benefits-of-half-precision-floats

**Comment by Randl** [comment]
@beru yeah the proposal exists http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0192r0.pdf hopefully will make it to next C++ standard. Implementing FP16 support without it is kinda optimistic I think -_-

**Comment by beru** [comment]
@Randl Yes, the most recent is probablly `P0192R1`.

Even if standard will adopt the short float type in the future, it will be hard to write programs without the support of Numerics library. And it takes some time for compilers and libraries vendors/implementors will ship their products. And this library is for `C++11` that means we can't rely on the new std for a while.

But boosting inference performance with FP16 on GPU has been common. I wonder how this library can support it...

**Comment by naibaf7** [comment]
@beru LibDNN is and OpenCL Caffe is scheduled to receive broad FP16 support in approx. September (I have a project planned for that), so maybe tiny-dnn can also profit from it after that.

**Comment by beru** [comment]
@naibaf7 
> LibDNN is and OpenCL Caffe is scheduled to receive broad FP16 support in approx. September (I have a project planned for that), so maybe tiny-dnn can also profit from it after that.

Sounds great! I've just checked Greentea LibDNN code but couldn't find `cl_khr_fp16` in
https://github.com/naibaf7/libdnn/blob/5a78b3e378cd042184eeb91fa6093875f9d12c5d/src/libdnn.cpp
So both of LibDNN and OpenCL Caffe are scheduled to have FP16 support in the future?
Or yet to be released LibDNN has already supported FP16?

**Comment by naibaf7** [comment]
@beru
Yes it's in development, however I currently lack good testing hardware to go further, such as AMD Vega.
Therefore the schedule for September.

**Comment by beru** [comment]
@naibaf7 It must be really cozy to stay warm with XX-XXX TFlops computers in winter season.

@Randl Let's support FP16 and integer types for storage type.
I think it should be dynamically configurable for flexibility, **not** via `typedef double|float float_t`.
It would be hard to implement all the kernels for basic types, so automatic type conversion mechanism will be needed.

**Comment by naibaf7** [comment]
@beru Yes even a mobile GPU with XXX GFlops makes for a good hand-warmer in winter *hint hint*
